edb/server/net_worker.py

```diff
@@ -0,0 +1,243 @@
+    HEAD = 'HEAD'
+    OPTIONS = 'OPTIONS'
+    PATCH = 'PATCH'
+
+
+@dataclass
+class ScheduledRequest:
+    id: str
+    method: HTTPMethod
+    url: str
+    body: typing.Optional[bytes]
+    headers: typing.Optional[dict[str, str]]
+
+    @staticmethod
+    def from_db_record(record: dict) -> ScheduledRequest:
+        id = record['id']
+        if not isinstance(id, str):
+            type_name = type(id).__name__
+            raise TypeError(f"Expected 'id' to be str, got {type_name}")
+
+        method = record['method']
+        if not isinstance(method, str):
+            type_name = type(method).__name__
+            raise TypeError(
+                f"Expected 'method' to be str, got {type_name}"
+            )
+        else:
+            try:
+                method = HTTPMethod(method)
+            except ValueError:
+                raise ValueError(f"Invalid HTTP method: {method}")
+
+        url = record['url']
+        if not isinstance(url, str):
+            type_name = type(url).__name__
+            raise TypeError(
+                f"Expected 'url' to be str, got {type_name}"
+            )
+
+        body = record.get('body')
+        if body is not None and not isinstance(body, bytes):
+            type_name = type(body).__name__
+            raise TypeError(
+                f"Expected 'body' to be bytes or None, got {type_name}"
+            )
+
+        maybe_headers = record.get('headers')
+        if maybe_headers is not None:
+            if not isinstance(maybe_headers, list):
+                type_name = type(maybe_headers).__name__
+                raise TypeError(
+                    f"Expected 'headers' to be list or None, got {type_name}"
+                )
+            if not all(
+                isinstance(name, str) and isinstance(value, str)
+                for name, value in maybe_headers
+            ):
+                raise TypeError(
+                    "Expected 'headers' to be a list of tuples of str"
+                )
+            else:
+                maybe_headers = typing.cast(
+                    list[tuple[str, str]], maybe_headers
+                )
+        headers = dict(maybe_headers) if maybe_headers else None
+
+        return ScheduledRequest(
+            id=id,
+            method=method,
+            url=url,
+            body=body,
+            headers=headers,
+        )
+
+
+async def http_send(request: ScheduledRequest) -> httpx.Response:
+    async with httpx.AsyncClient() as client:
+        response = await client.request(
+            request.method,
+            request.url,
+            content=request.body,
+            headers=request.headers,
+        )
+        return response
+
+
+async def handle_request(
+    db: dbview.Database, request: ScheduledRequest
+) -> None:
+    try:
+        response = await http_send(request)
+        request_state = 'Completed'
+        response_status = response.status_code
+        response_body = response.content
```

Comment:
`read` isn't awaitable? It seems like it returns `bytes` here? Maybe the change is:

```suggestion
        response_body = response.read()
```

The documentation doesn't seem to give a really clear picture on how to do what we're trying to do here which is save the response bytes without any extra decoding. Although... Maybe we _want_ it decoded (to account for compression) and the re-encode it so that we're saving something that can be easily to deal with on the consumer's end?

Reply:
Oh sorry it should be `await response.aread()`.

> without any extra decoding

I suppose you meant the textual codecs (UTF-8, ISO-8859-1, etc.) instead of the protocol codes (gzip, deflate, etc.), right? There is the `response.aiter_text()` but we'll then need to handle the decoding errors - coming back to the previous question of storing partial responses. I'd prefer we just keep the `bytes` (only decoding the gzip/deflate/... stuff).

Or maybe we can try to decode it into `str` and - like you said - store encoded bytes with confidence (like with an additional optional field on the `Response` indicating the confirmed text encoding, or even always re-encoded as UTF-8), or just store the raw bytes if we failed to decode.

Reply:
> I suppose you meant the textual codecs (UTF-8, ISO-8859-1, etc.) instead of the protocol codes (gzip, deflate, etc.), right?

Ahh, I mostly meant compression, not text encodings, but it looks like `response.content` already handles decompression:

https://www.python-httpx.org/quickstart/#binary-response-content

> Any gzip and deflate HTTP response encodings will automatically be decoded for you. If brotlipy is installed, then the brotli response encoding will be supported. If zstandard is installed, then zstd response encodings will also be supported.

> Oh sorry it should be await response.aread().

👍 

> Or maybe we can try to decode it into str and - like you said - store encoded bytes with confidence (like with an additional optional field on the Response indicating the confirmed text encoding, or even always re-encoded as UTF-8), or just store the raw bytes if we failed to decode.

The main problem I was worried about was having consumers have to read a bunch of headers to know how to decompress the bytes into the encoded text, but since that's not really the issue here, I'm personally fine with just storing the raw bytes, and having the consumer decode themselves in what is most likely utf-8. Although, maybe that's relatively hostile since we imagine that these responses will typically only be read at troubleshooting time, and typically via the UI which would be a subpar experience.

On the other hand, I _really_ don't want to make this part of the module super complicated and the simplest thing to actually store is the raw bytes. @1st1 any thoughts here?